backends/optimized/ceed-opt-basis.c

```diff
@@ -57,108 +54,106 @@ static int CeedBasisApply_Opt(CeedBasis basis, CeedInt nelem,
   int ierr;
   const CeedInt dim = basis->dim;
   const CeedInt ncomp = basis->ncomp;
-  const CeedInt nqpt = ncomp*CeedPowInt(basis->Q1d, dim);
+  const CeedInt nqpt = CeedPowInt(basis->Q1d, dim);
   const CeedInt add = (tmode == CEED_TRANSPOSE);
+  const CeedInt blksize = 8;
 
-  if (nelem != 1)
+  if ((nelem != 1) && (nelem != blksize))
     return CeedError(basis->ceed, 1,
-                     "This backend does not support BasisApply for multiple elements");
+                     "This backend does not support BasisApply for %d elements", nelem);
 
   if (tmode == CEED_TRANSPOSE) {
-    const CeedInt vsize = ncomp*CeedPowInt(basis->P1d, dim);
+    const CeedInt vsize = nelem*ncomp*CeedPowInt(basis->P1d, dim);
     for (CeedInt i = 0; i < vsize; i++)
-      v[i] = (CeedScalar) 0;
+      v[i] = (CeedScalar) 0.0;
   }
-  if (emode & CEED_EVAL_INTERP) {
+  if (emode == CEED_EVAL_INTERP) {
     CeedInt P = basis->P1d, Q = basis->Q1d;
     if (tmode == CEED_TRANSPOSE) {
       P = basis->Q1d; Q = basis->P1d;
     }
-    CeedInt pre = ncomp*CeedPowInt(P, dim-1), post = 1;
-    CeedScalar tmp[2][ncomp*Q*CeedPowInt(P>Q?P:Q, dim-1)];
+    CeedInt pre = ncomp*CeedPowInt(P, dim-1), post = nelem;
+    CeedScalar tmp[2][nelem*ncomp*Q*CeedPowInt(P>Q?P:Q, dim-1)];
     for (CeedInt d=0; d<dim; d++) {
-      ierr = CeedTensorContract_Opt(basis->ceed, pre, P, post, Q, basis->interp1d,
-                                    tmode, add&&(d==dim-1),
+      ierr = CeedTensorContract_Opt(basis->ceed, pre, P, post, Q,
+                                    basis->interp1d, tmode, add&&(d==dim-1),
                                     d==0?u:tmp[d%2], d==dim-1?v:tmp[(d+1)%2]);
       CeedChk(ierr);
       pre /= P;
       post *= Q;
     }
-    if (tmode == CEED_NOTRANSPOSE) {
-      v += nqpt;
-    } else {
-      u += nqpt;
-    }
-  }
-  if (emode & CEED_EVAL_GRAD) {
-    CeedInt P = basis->P1d, Q = basis->Q1d;
+  } else if (emode == CEED_EVAL_GRAD) {
     // In CEED_NOTRANSPOSE mode:
-    // u is (P^dim x nc), column-major layout (nc = ncomp)
-    // v is (Q^dim x nc x dim), column-major layout (nc = ncomp)
+    // u is (P^dim x nc) x nelem, column-major layout (nc = ncomp)
+    // v is (Q^dim x nc x dim) x nelem, column-major layout (nc = ncomp)
```

Comment:
Except the `nelem` part is the contiguous direction. We should adopt a consistent convention about expressing shapes. I would propose the C-inspired `[dim, ncomp, Q^dim, nelem]`. We can do that as part of a sanitation pass later.

Reply:
I'm good to merge; I'll start the sanitation pass branch today.

Can you put those compile options you were using in your branch please?

Reply:
ce77c5bc200beacfeecb101312aa5f3112587880

Reply:
Do you want to cherry-pick cf12fa7df7c3e65edbafc900d2c138497e2eec99 into your branch now?

Reply:
Updated